### ISSUE TITLE
Launcher failing to update mods correctly, causing load errors

### ISSUE THREAD

**Issue by chaosvolt** [remyroy/CDDA-Game-Launcher#206]
https://cdn.discordapp.com/attachments/252368987210514433/479797999048851457/mj.PNG

Confirmed to affects Arcana and Magic Items, Cata++, probably PK's Rebalance. Others probable but unsure.

For whatever reason, people updating their mods via the launcher are still getting errors caused by https://github.com/CleverRaven/Cataclysm-DDA/pull/24183, despite the master branches of affected mods having since been updated to fix that.

It's gotten to the point where I've had to tell several people to update their mods manually instead of using the launcher, and I STILL get complaints about a bug I've already fixed.

**Comment by remyroy** [comment]
What is the solution here? Are there new updated mods that should replace mods included in the launcher repository?

**Comment by chaosvolt** [comment]
Not that I know of, I'm not even sure what the cause is. All I know is I've had to constantly tell people not to use the launcher to update Arcana or Cata++, because people are still, to this day, getting errors like missing threshold_mut when they use the launcher to do so.

Since they're able to grab the outdated version of them, that at least suggests they you've got the right repos, and far as I know the repos haven't changed. But something here is screwed up for that to still be happening.

**Comment by remyroy** [comment]
Should I just remove Arcana and Cata++ from the launcher repository so they can't install those mods?

**Comment by chaosvolt** [comment]
New potential thing to test as well: recently Cata++ updated to fix the load errors caused by having redundant "revert to null item" properties specified.

**Comment by chaosvolt** [comment]
So I've been discussing things with others more familiar with the launcher and I have a question. Does the launcher automatically update mods at all? Because evidently the UI has no update button, making it seem as though the only way to update is via deleting and re-adding the mod in the modlist.

If it DOES automatically update mods then my issue here is still relevant, as something there absolutely needs to be fixed, because I am still handling "oh hey updating the mod manually fixed it" problems every single day on the discord server.

If it DOESN'T automatically update mods and you have to do it manually on the launcher, then it's most likely half user error and half bad design, because having no indication that the user is expected to update the mods themselves means they likely won't.

**Comment by chaosvolt** [comment]
Might be good to warn users of that in the mods tab then.

**Comment by KurzedMetal** [comment]
@remyroy I've read in the CDDA Discord that some players get errors because the Launcher never updates the mods after the first installation because some mods have to "catch up" with main CDDA code to actually work. ChaosVolt also pointed me to this bug.

I was thinking, maybe I could add an "update" button that does an "uninstall & resintall" in one go without doing any kind of version management (which would be pretty much impossible to implement in the way the current way modders host or distribute stuff and without coordination)

It would basically blindly redownload and install latest version without caring if it was the same as before (granted the devs of the mod actually host the latest version in a specific URL).

I could even add some button below / to the right of "Update game" named "Update all mods", so the player can do it in one go (or even a setting to always update mods when CDDA is updated).

Would you be OK to add such a feature if I implement it?
Did you have any plans or ideas to implement something like that?

**Comment by remyroy** [comment]
It's almost impossible to tell when or if a mod needs to be updated. The *official* solution is to have the mod included in the base game which does not suit some mod authors or it does not suit the game authors/maintainers for some mods.

I don't have a plan for providing any mod update feature. I think it could be interesting for some users, but I also think it could be detrimental for some. The case I'm thinking about here is a user who has not updated his game for like 3 months who sees this *Update all mods* button, update his mods thinking he is going to get new stuff in and break his game because the mods are now made for a more recent version of the game. I'm sure there are plenty more cases where having or not having this feature break stuff.

I feel like adding such a feature is going to be a pain to support and that it's going to bring a huge number of support requests from users. I feel like this is going to be a fragile feature.

Ideally if we could test specific versions of a mod against specific versions of the game and compile the results somewhere, we could offer more insight to users such as warnings before installing if a mod is incompatible with the game or automatically updating mods when they become incompatible with the game. Unfortunately, there is almost nothing that provides this kind of information and the game authors/maintainers made it clear they don't want to go that route.

So yeah, there is no perfect solution here. People are going to be mad if we don't have it and people are going to be mad if we have it. I'm inclined not to add such feature.

**Comment by KurzedMetal** [comment]
Agree, there's no easy way to do a version check (specially not without downloading the mod first), that's why I was suggesting a "dumb" version of an "updater" (calling it "reinstaller" would be more appropriate), which just remove and install latest zip file.

> user who has not updated his game for like 3 months who sees this Update all mods button, update his mods thinking he is going to get new stuff in and break

That could be easily fixable by disabling the "Update all mods" button if CDDA is not on latest version.

>  if we could test specific versions of a mod against specific versions of the game and compile the results somewhere

I was actually trying to think/design a solution that would require no maintenance.

**Comment by remyroy** [comment]
The *update all mods* could work but it could also fail even if we make sure the player is on the latest experimental version. For instance, if the game just broke their mod API (whatever that might be), some mod author updated their mods for this new API and some did not update their mods for this new API. Updating all mods would work for those mods that are now using the new API and it would break for those that do not.

I previously suggested the game authors/maintainers to do some kind of mod API versioning. It could have been a single value like version 1, 2, 3, etc and block or warn about mods that do not target the same version in the current game or in the mod selection screen but that was promptly rejected. It would have greatly helped with most of these issues and we could have inspected the game and the mods to have a matching mod API version.

In the current environment, this is just a hard issue that we can't do much about in my humble opinion.

**Comment by KurzedMetal** [comment]
> some mod author updated their mods for this new API and some did not update their mods for this new API. Updating all mods would work for those mods that are now using the new API and it would break for those that do not.

But in that situation the mod will break anyways if the player updates CDDA. With or without updating the mod.
That's not responsibility of the launcher. Specially if there's no versioning info about mods/game anywhere. If players detect something like that they can still rollback the updates (or pick a previous CDDA version) to get a working setup or play without that mod activated, or inform the mod maintainers.

Even if it we don't have a designed mod versioning to do a better handling of updates, that doesn't mean the launcher cannot try doing it. And I think an "unreliable update" is better than no update, because it will work better for those mod maintainers that need to do changes on their mods and actually do it. Also, players end up going it manually anyways. Basically, what I am sayng is: I'd pick the lesser evil :P

Anyways, I do understand and agree on all your points tho, they are all very valid and I thought about them myself too.

**Comment by remyroy** [comment]
It's true that right now, a game update with a breaking mod API update is likely to break stuff no matter if the mod author updated his mods accordingly or if the player updates his mods or not.

We could initially add that flag where we know that a specific mod is updated frequently and provide that update/reinstall feature for those mods.
